src/Cryptol/Backend/SeqMap.hs

```diff
@@ -173,13 +176,21 @@ memoMap _sym _sz x@(MemoSeqMap{}) = pure x
 
 memoMap sym sz x = do
   stk <- sGetCallStack sym
-  cache <- liftIO $ newIORef $ Map.empty
-  evalRef <- liftIO $ newIORef $ eval stk
-  return (MemoSeqMap sz cache evalRef)
+  cache <- liftIO $ newIORef (Map.empty, eval stk)
+  return (MemoSeqMap sz cache evalPanic)
 
   where
     eval stk i = sWithCallStack sym stk (lookupSeqMap x i)
 
+    -- Not 100% sure that we actually need to do the bounds check here,
+    -- or if we are assuming that the argument would be in bounds, but
+    -- it doesn't really hurt to do it, as if we *did* already do the check
+    -- this code should never run (unless we messed up something).
+    evalPanic i = case sz of
+                    Nat sz' | i < 0 || i >= sz' -> invalidIndex sym i
+                    _ -> panic "lookupSeqMap"
```

Comment:
Nit pick: this function is no longer located in `lookupSeqMap`, so it would be worth changing the string used in the call to `panic` here.